### DISCUSSION TITLE
convert png to avif lossless, but not lossless

### DISCUSSION THREAD

**Discussion by oldshensheep** [ImageMagick/ImageMagick#6591] [Help]
### ImageMagick version

ImageMagick 7.1.1-15 Q16-HDRI x64 a0a5f3d:20230730

### Operating system, version and so on

Windows 11

### Description

0.png
![0](https://github.com/ImageMagick/ImageMagick/assets/29306733/16dbddf2-d984-4b92-bc48-8779be28b93a)
`-quality 100` is lossless https://github.com/ImageMagick/ImageMagick/discussions/3076 
```
> magick.exe -quality 100 0.png 0.avif
> compare.exe -metric ssim 0.png 0.avif out.png
0.996046 (except 1)
```
But if I use WebP it works as expected.
```
> magick.exe -quality 100 -define webp:lossless=true 0.png 0.webp
> compare.exe -metric ssim 0.png 0.webp out.png
1
```
I'm searching and found this https://github.com/AOMediaCodec/av1-avif/issues/111#issuecomment-717710961
>Not only that, but AVIF lossless isn't even true lossless, since it requires the image be converted to yuv4:4:4 or yuv4:2:0, which means if your image is originally RGB, you're incurring a lossy conversion just by changing from RGB to YUV. I've heard of ways in which the image can be losslessly converted between RGB and YUV, but that is unorthodox and I do not recommend it.

**First Question**: Is that true? If an Image is in RGB converted to avif lossless, but it is actually a lossy conversion?

I look up into code
https://github.com/ImageMagick/ImageMagick/blob/1bca2e1a28bbc36143d4d39d7ead62928be04db1/coders/heic.c#L1225

https://github.com/ImageMagick/ImageMagick/blob/1bca2e1a28bbc36143d4d39d7ead62928be04db1/coders/heic.c#L1246
So if quality >=100 that's lossless=true, and if lossless=true chroma=heif_chroma_444 right?
**Second Question** :If so, why exiftool shows `Chroma Format                   : YUV 4:2:0` ?
```
MIME Type                       : image/avif
Major Brand                     : AV1 Image File Format (.AVIF)
Minor Version                   : 0.0.0
Compatible Brands               : avif, mif1, miaf
Handler Type                    : Picture
Primary Item Reference          : 1
Color Profiles                  : nclx
Color Primaries                 : Unspecified
Transfer Characteristics        : Unspecified
Matrix Coefficients             : BT.601
AV1 Configuration Version       : 1
Chroma Format                   : YUV 4:2:0
Chroma Sample Position          : Unknown
Image Width                     : 460
Image Height                    : 460
Image Spatial Extent            : 460x460
Image Pixel Depth               : 8 8 8
Media Data Size                 : 65041
Media Data Offset               : 270
Image Size                      : 460x460
Megapixels                      : 0.212
```

**Comment by remino** [comment]
Came on here to say I just tried converting a PNG to AVIF with `-quality=100` myself and I’m also getting an actual lossy AVIF, not lossless.

**Reply by dlemstra** [reply]
Can you reproduce this with the latest version of ImageMagick?

**Reply by remino** [reply]
Sure. I went on set up a PoC repo to show what’s happening on my end:

https://github.com/remino/magick-lossless-avif-issue

What I get with ImageMagick 7.1.1-29 are the following two images for comparison, the PNG on the left below, and AVIF on the right. (GitHub may recompress the images and add more artefacts.)

![cropped-png](https://github.com/ImageMagick/ImageMagick/assets/29999/5cf23d93-9c40-45d6-8d99-2c3def380403)
![cropped-avif](https://github.com/ImageMagick/ImageMagick/assets/29999/fb6111f2-70fb-4732-9b24-f4aa1eb6a96a)

In the AVIF format, it looks like some kind of emboss is added to the red shape. Now I can’t tell if it’s because of ImageMagick or because of the AVIF format itself. But I imagined that if it’s lossless, it should be the same output.

**Reply by dlemstra** [reply]
This is probably happening in the AVIF encoder and there is only a really small difference between those two images. Probably due to rounding errors.

**Comment by darkwasthenight** [comment]
I had a similar issue. after playing around with the compression settings i found a good one for me.
i'm  going with 80% as a result, the loss is almost invisible, but the size is very impressive as compared to a similar quality in webp
here is an example of a [skateboard](https://www.skateshop24.de/avif/thumbnail/5e/d8/ef/1737757882/little-boards-rocket-01-jpg_1920x1920.avif) with just 16KB

**Comment by kolAflash** [comment]
This issue might be the answer needed here:  
* https://github.com/ImageMagick/ImageMagick/issues/8051#issuecomment-2992773501
> AVIF is not entirely lossless. It stores the underlying YUV data losslessly, but it still converts from RGB to YUV with some loss of precision. This is why e.g. GIMP calls it "almost lossless".

**Comment by fmw42** [comment]
FYI:

`magick.exe -quality 100 0.png 0.avif`

you should read the png file before -quality, assuming -quality is for the avif.  But if I recall, avif is not always lossless by setting -quality 100.

**Reply by lilydjwg** [reply]
I got an error with `-quality 100`. What went wrong?

```
magick: Encoder plugin generated an error: Unsupported parameter: AOM encoder error: Invalid parameter - Only --enable_chroma_deltaq=0 can be used with --lossless=1. (8.2005) `b.avif' @ error/heic.c/IsHEIFSuccess/206.
```

**Comment by fmw42** [comment]
`magick.exe -quality 100 0.png 0.avif`

On IM 7, you need to read the input before other settings and commands. IM 7 command line working in series. So it you put the input at the end, the settings and commands do not know what to process.  

`magick.exe 0.png -quality 100 0.avif`

That command works fine for me on IM 7.1.2.26 Mac OSX Ventura

`magick 0.png -quality 100 0.avif`

[Archive.zip](https://github.com/user-attachments/files/29254705/Archive.zip)

**Reply by lilydjwg** [reply]
I'm using 7.1.2.25 on Linux. I tried to put `-quality 100` before input and it still gives the error. However it works with webp output.

I checked your attachment. It fills the last information I was wanting to know.

(I was exploring if I could use avif or webp instead of png for screenshots. The avif file is still yuv420p and not pixel-identical, which means that it is not suitable for high quality screenshots. With `avifenc -q 100` I can get a yuv444p output and it looks the same despite still not pixel-identical, so it can be used for most screenshot purposes for me. `avifenc` has a `-l` option that outputs pixel-identical results, but with larger file sizes.)

**Comment by fmw42** [comment]
**I said try IM 7.1.2.26, _not_ IM 7.1.2.25**.  See if that works for you. It worked for me.

**Reply by lilydjwg** [reply]
7.1.2.26 has just arrived to me. The error is the same.

**Comment by q3cpma** [comment]
`magick SRC.png -quality 100 -define heic:chroma=444 -define heic:cicp=1/13/0/1 DST.avif` worked here. There are some recent PRs that seem related to the matter (https://github.com/ImageMagick/ImageMagick/pull/8824 and https://github.com/ImageMagick/ImageMagick/pull/8828) but I don't think these fix this issue.

**Reply by lilydjwg** [reply]
Thanks, your command works and produces a clear result for screenshots. Unfortunately the file size is larger than PNG.

**Reply by q3cpma** [reply]
Did you try with `-define heic:speed=0` too? But yeah, AVIF isn't a good lossless format.

**Reply by lilydjwg** [reply]
That option took me 54 seconds to transcode a 499K PNG into a 564K AVIF. So yeah, it is not good at screenshots.

**Comment by pinballmachine** [comment]
AVIF can be used in RGB mode, without conversion to YUV. For example [cavif](https://github.com/kornelski/cavif-rs) supports it. However, `cavif` does not support a lossless mode. The reference encoder [avifenc](https://github.com/AOMediaCodec/libavif) supports lossless mode, but not RGB mode, so not truly lossless I guess.
I don't know which encoder ImageMagick uses for AVIF. If it's `libheif`, that can be built with various encoders under the hood (`libaom`, `dav1d`, `svt-av1`, or `rav1e`), and they might not be up-to-date.

Probably you will find that creating lossless AVIF images will (almost?) always result in bigger files than well optimized PNGs.*

The best-looking, most-optimized and most-predictable (size-wise) results you will get with `avifenc` in 10-bit mode. That will prevent rounding errors during the conversion from RGB to YUV. It used to be that you needed the `tune=iq` switch, but AFAIK that is now default in the latest version. For me, `avifenc` on quality 80 is visually lossless. You can leave it at the default speed (=speed 6). It is very quick, and going slower barely gives any gains.
`avifenc.exe -d 10 -q 80 input.png output.avif` will produce stellar results at great speed.

Off-topic, but I find it funny that if you take a look at a lossy [JPEG-XL vs AVIF example](https://jpegxl.info/) I actually think the AVIF looks better. JPEG-XL does have a lossless mode you could try though.

* If you want to go nuts and squeeze out the last few bytes you can recompress your PNG with `pngcrush`, `optipng`, `pngout`, `oxipng` and `pngwolf-zopfli` in that order.
